sycl/source/detail/platform_impl.cpp

```diff
@@ -26,6 +27,50 @@ namespace detail {
 
 using PlatformImplPtr = std::shared_ptr<platform_impl>;
 
+// Return true if the given device 'Dev' matches with any filter
+bool isForcedDevice(detail::device_filter_list *FilterList,
+                    const plugin &Plugin, const RT::PiDevice &Dev,
+                    unsigned int Index) {
+  backend Backend = Plugin.getBackend();
+  info::device_type Type;
+  Plugin.call<PiApiKind::piDeviceGetInfo>(
+      Dev, pi::cast<RT::PiDeviceInfo>(info::device::device_type), sizeof(Type),
+      &Type, nullptr);
+
+  // According to SYCL spec host device must be always available
+  if (Type == info::device_type::host)
+    return true;
+
+  for (const detail::device_filter &Filter : FilterList->get()) {
+    if ((Filter.Backend == Backend || Filter.Backend == backend::all) &&
+        (Filter.DeviceType == Type ||
+         Filter.DeviceType == info::device_type::all)) {
+      if (!Filter.HasDeviceNum ||
+          static_cast<unsigned int>(Filter.DeviceNum) == Index)
+        return true;
+    }
+  }
+  return false;
+}
+
+static void applyDeviceFilter(vector_class<RT::PiDevice> &PiDevices,
+                              const plugin &Plugin) {
+  detail::device_filter_list *FilterList =
+      detail::SYCLConfig<detail::SYCL_DEVICE_FILTER>::get();
+  if (!FilterList)
+    return;
+
+  unsigned int InsertIDx = 0;
+  for (unsigned int i = 0; i < PiDevices.size(); i++) {
+    if (isForcedDevice(FilterList, Plugin, PiDevices[i], i)) {
+      if (InsertIDx != i)
+        PiDevices[InsertIDx] = PiDevices[i];
+      InsertIDx++;
+    }
```

Comment:
I think it is better to add 'HOST' device separately from the forced devices rather than counting the host device as the default forced device.
The reason is that we really need to give a bonus point that matches with user-specified device than the automatically included 'host' device.

Reply:
done